File: backend/utils/json_column_storage.py
"""
Persist JSON array columns as NULL when empty (never literal ``[]`` in SQLite/Postgres TEXT).
"""
from __future__ import annotations

import json
from typing import Any


def nullable_json_array_text(value: Any) -> str | None:
    """Return JSON array text for storage, or NULL when the array is empty."""
    if value is None:
        return None
    if isinstance(value, list):
        return json.dumps(value, ensure_ascii=False) if value else None
    if isinstance(value, str):
        s = value.strip()
        if not s or s in ("[]", "null"):
            return None
        try:
            parsed = json.loads(s)
            if isinstance(parsed, list):
                return s if parsed else None
        except (json.JSONDecodeError, TypeError):
            pass
        return s
    return None

File: backend/scanner/json_column_storage.js
/**
 * Persist JSON array columns as NULL when empty (never literal "[]" in TEXT columns).
 */

function nullableJsonArrayText(value) {
  if (value == null) return null;
  if (Array.isArray(value)) return value.length ? JSON.stringify(value) : null;
  if (typeof value === "string") {
    const s = value.trim();
    if (!s || s === "[]" || s === "null") return null;
    try {
      const parsed = JSON.parse(s);
      if (Array.isArray(parsed)) return parsed.length ? s : null;
    } catch (_e) {
      return s;
    }
    return s;
  }
  return null;
}

module.exports = { nullableJsonArrayText };
